control-plane/agents/core/src/core/states.rs

```diff
@@ -0,0 +1,88 @@
+pub(crate) struct ResourceStatesLocked(Arc<RwLock<ResourceStates>>);
+
+impl ResourceStatesLocked {
+    pub(crate) fn new() -> Self {
+        ResourceStatesLocked::default()
+    }
+}
+
+impl Deref for ResourceStatesLocked {
+    type Target = Arc<RwLock<ResourceStates>>;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+/// Resource States
+#[derive(Default, Debug)]
+pub(crate) struct ResourceStates {
+    /// Todo: Add runtime state information for nodes.
+    pub(crate) nexuses: HashMap<NexusId, Arc<Mutex<NexusState>>>,
+    pub(crate) pools: HashMap<PoolId, Arc<Mutex<PoolState>>>,
+    pub(crate) replicas: HashMap<ReplicaId, Arc<Mutex<ReplicaState>>>,
+}
+
+impl ResourceStates {
+    /// Update the pool states with the latest runtime information from Mayastor.
+    pub(crate) async fn update_pools(&mut self, pools: &[Pool]) {
+        for pool in pools {
+            let pool_state = PoolState::from(pool);
+            // If an entry exists for the pool, update it, otherwise add a new entry.
+            match self.pools.get(&pool.id) {
+                Some(locked_state) => {
+                    let mut state = locked_state.lock().await;
+                    *state = pool_state;
+                }
+                None => {
+                    self.pools
+                        .insert(pool.id.clone(), Arc::new(Mutex::new(pool_state)));
+                }
+            }
+        }
+    }
+
+    /// Update the replica states with the latest runtime information from Mayastor.
+    pub(crate) async fn update_replicas(&mut self, replicas: &[Replica]) {
+        for replica in replicas {
+            let replica_state = ReplicaState::from(replica);
+            // If an entry exists for the replica, update it, otherwise add a new entry.
+            match self.replicas.get(&replica.uuid) {
+                Some(locked_state) => {
+                    let mut state = locked_state.lock().await;
+                    *state = replica_state;
+                }
+                None => {
+                    self.replicas
+                        .insert(replica.uuid.clone(), Arc::new(Mutex::new(replica_state)));
+                }
+            }
+        }
+    }
+
+    /// Update the nexus states with the latest runtime information from Mayastor.
+    pub(crate) async fn update_nexuses(&mut self, nexuses: &[Nexus]) {
+        for nexus in nexuses {
+            let nexus_state = NexusState::from(nexus);
+            // If an entry exists for the nexus, update it, otherwise add a new entry.
+            match self.nexuses.get(&nexus.uuid) {
+                Some(locked_state) => {
+                    let mut state = locked_state.lock().await;
+                    *state = nexus_state;
+                }
+                None => {
+                    self.nexuses
+                        .insert(nexus.uuid.clone(), Arc::new(Mutex::new(nexus_state)));
+                }
+            }
+        }
+    }
+}
```

Comment:
nit; There are 3 things that all do the same... seems like it can be made generic of the resources

Reply:
I've made a generic function for these.